File: lambda_counter.js
const { DynamoDBClient, GetItemCommand, PutItemCommand } = require("@aws-sdk/client-dynamodb");
const { SecretsManagerClient, GetSecretValueCommand } = require("@aws-sdk/client-secrets-manager");

const dynamoClient = new DynamoDBClient({});
const secretsClient = new SecretsManagerClient({ region: "us-east-1" });
const tableName = "db-visit-count";
const secretName = "test/test2";

exports.lambdaHandler = async (event, context) => {
    let body;
    let statusCode = 200;
    const headers = { "Content-Type": "application/json" };
    let visitCount = 0;

    try {
        const routeKey = event.routeKey;

        if (routeKey === "GET /items/{user}") {
            const user = event.pathParameters.user;

            // Get current visit count
            const response = await dynamoClient.send(new GetItemCommand({
                TableName: tableName,
                Key: { "user": { S: user } }
            }));

            // Extract count
            if (response.Item) {
                visitCount = parseInt(response.Item.count?.N || "0", 10);
            }

            // Increment count
            visitCount += 1;
            const message = `Visit count is ${visitCount}`;

            // Reconstruct Item
            const item = {
                "user": { S: user },
                "count": { N: visitCount.toString() }
            };

            // Write visit count back to DynamoDB
            await dynamoClient.send(new PutItemCommand({ TableName: tableName, Item: item }));

            body = JSON.stringify(message);
        } else if (routeKey === "GET /secret") {
            // Retrieve secret from AWS Secrets Manager
            const secretResponse = await secretsClient.send(new GetSecretValueCommand({
                SecretId: secretName,
                VersionStage: "AWSCURRENT"
            }));

            body = secretResponse.SecretString;
        } else {
            throw new Error(`Unsupported route: ${routeKey}`);
        }
    } catch (error) {
        statusCode = 400;
        body = error.message;
    }

    return {
        statusCode,
        body,
        headers
    };
};

File: lambda_counter.py
import json
import boto3
import os
dynamodb = boto3.client("dynamodb")
table_name = "db-visit-count"

def lambda_handler(event, context):
    body = None
    status_code = 200
    headers = {
        "Content-Type": "application/json"
    }
    visit_count: int = 0
    try:
        route_key = event["routeKey"]
        
        if route_key == "GET /items/{user}":

            #get current visit count
            response = dynamodb.get_item(
                TableName=table_name,
                Key={
                    "user": {"S": event["pathParameters"]["user"]}
                }
            )

            #extract count
            if "Item" in response:
                visit_count = int(response.get('Item',{}).get('count',{}).get('N'))

            #increment count
            visit_count += 1
            message = f"Visit count is {visit_count}"
            
            #reconstruct Item 
            item = {"user": {"S": event["pathParameters"]["user"]}, "count": {} }
            item['count']['N'] = str(visit_count)

            #write visit-count back to dynamodb
            dynamodb.put_item(
                TableName=table_name,
                Item = item
            )
            body = message
        
        else:
            raise Exception(f"Unsupported route: {route_key}")
    except Exception as e:
        status_code = 400
        body = str(e)
    finally:
        body = json.dumps(body)
    return {
        "statusCode": status_code,
        "body": body,
        "headers": headers
    }


    